bin/runtime-common/src/refund_relayer_extension.rs

```diff
@@ -268,118 +309,55 @@ where
-		// this is the only relevant line of code for the `pre_dispatch`
-		//
-		// we're not calling `validato` from `pre_dispatch` directly because of performance
-		// reasons, so if you're adding some code that may fail here, please check if it needs
-		// to be added to the `pre_dispatch` as well
-		let parsed_call = self.parse_and_check_for_obsolete_call(call)?;
-
-		// the following code just plays with transaction priority and never returns an error
-		let mut valid_transaction = ValidTransactionBuilder::default();
-		if let Some(parsed_call) = parsed_call {
-			// we give delivery transactions some boost, that depends on number of messages inside
-			let messages_call_info = parsed_call.messages_call_info();
-			if let MessagesCallInfo::ReceiveMessagesProof(_) = messages_call_info {
-				// compute total number of messages in transaction
-				let bundled_messages = messages_call_info.bundled_messages();
-
-				// a quick check to avoid invalid high-priority transactions
-				if bundled_messages <= Runtime::MaxUnconfirmedMessagesAtInboundLane::get() {
-					let priority_boost = crate::priority_calculator::compute_priority_boost::<
-						Priority,
-					>(bundled_messages);
-					valid_transaction = valid_transaction.priority(priority_boost);
-				}
-			}
-		}
+	/// Given post-dispatch information, analyze the outcome of relayer call and return
+	/// actions that ened to be performed on relayer account.
+	fn analyze_call_result(
+		pre: Option<Option<PreDispatchData<Runtime::AccountId>>>,
+		info: &DispatchInfo,
+		post_info: &PostDispatchInfo,
+		len: usize,
+		result: &DispatchResult,
+	) -> RelayerAccountAction<AccountIdOf<Runtime>, Runtime::Reward> {
+		let mut extra_weight = Weight::zero();
+		let mut extra_size = 0;
 
-		valid_transaction.build()
-	}
+		// We don't refund anything for transactions that we don't support.
+		let (relayer, call_info) = match pre {
+			Some(Some(pre)) => (pre.relayer, pre.call_info),
+			_ => return RelayerAccountAction::None,
+		};
 
-	fn pre_dispatch(
-		self,
-		who: &Self::AccountId,
-		call: &Self::Call,
-		_info: &DispatchInfoOf<Self::Call>,
-		_len: usize,
-	) -> Result<Self::Pre, TransactionValidityError> {
-		// this is a relevant piece of `validate` that we need here (in `pre_dispatch`)
-		let parsed_call = self.parse_and_check_for_obsolete_call(call)?;
+		// now we know that the relayer either needs to be rewarded, or slashed
+		// => let's prepare the correspondent account that pays reward/receives slashed amount
+		let reward_account_params = RewardsAccountParams::new(
+			Msgs::Id::get(),
+			Runtime::BridgedChainId::get(),
+			if call_info.is_message_delivery_call() {
+				RewardsAccountOwner::ThisChain
+			} else {
+				RewardsAccountOwner::BridgedChain
+			},
+		);
+		let is_message_delivery_call = call_info.is_message_delivery_call();
+		let slash_relayer_if_delivery_result = is_message_delivery_call
+			.then(|| RelayerAccountAction::Slash(relayer.clone(), reward_account_params))
```

Comment:
The priority is boosted only if relayer is registered, right? And the other part is controversial - if registered relayer submits incorrect transaction with more messages than the call accepts, shall we slash? Even if priority is not boosted? He's clearly doing something wrong. WDYT?

Reply:
Agree. But still feels strange to slash if the priority wasn't boosted. Maybe if `bundled_messages > Runtime::MaxUnconfirmedMessagesAtInboundLane::get()`, `validate()` should fail ? Wouldn't the transaction be invalid ?

Reply:
Also unregistered relayers could submit such transactions, but they wouldn't get slashed. Feels unfair.

Reply:
> Agree. But still feels strange to slash if the priority wasn't boosted. Maybe if bundled_messages > Runtime::MaxUnconfirmedMessagesAtInboundLane::get(), validate() should fail ? Wouldn't the transaction be invalid ?

We can reject transaction from signed extension, but then it'll be free for relayer. If we fail it in the call code, he'll pay the full fee.

> Also unregistered relayers could submit such transactions, but they wouldn't get slashed. Feels unfair.

Yep, that's a valid point. Unfair indeed. I'll update the code

Reply:
> We can reject transaction from signed extension, but then it'll be free for relayer. If we fail it in the call code, he'll pay the full fee.

Just a few more words: if submitter is doing something definitely wrong, then he should pay for that. We are rejecting obsolete transactions from signed extensions only because several honest relayer may submit valid transactions => one of them will be the winner and all others - just lose their money and waste block time, even though they have submitted valid transactions. Once we'll have some protocol for handling concurrent relayers, this extension will gone

Reply:
Sorry, I don't get the idea. The relayer should prepare correct transaction and submit it. The only error that the honest relayer can made (and he can't do nothing about it right now) is that some other transaction may be mined before this transaction, making his transaction obsolete. That's why we return `Stale` from our signed extension `pre_dispatch`. All other issues are result of relayer misbehavior, so we slash if he has a stake.

Reply:
Ok, just wanted to double check, make sure I read everything correctly 👍🏻